GH-37876: [Format] Add list-view specification to arrow format (#37877)
apache/arrow: docs/source/format/Columnar.rst

Nested arrays additionally have a sequence of one or more sets of
these items, called the **child arrays**.

> Each logical data type has one or more well-defined physical layouts. Here

Reviewer: I would keep the singular. There is no disjunction in Arrow (unlike Parquet) between "logical" data type and physical layout. ListView and StringView are simply distinct types.

Reviewer: I definitely agree with that. The spec was often confusing to me at the start.

          |-------------------------------|-----------------------|
          | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 | unspecified (padding) |

ListView Layout
~~~~~~~~~~~~~~~

The ListView layout is defined by three buffers instead of just two:
a validity bitmap, an offsets buffer, and an additional sizes buffer.

> The sizes have the same bit width as the offsets and both 32-bit and 64-bit
> signed integer options are supported. Like in the List layout, the offsets
> reference the child array.
> 
> Rather then inferring list lengths from the offsets, the sizes buffer
> stores the length of each list in the array. This in turn allows offsets to be

Reviewer 2: ```suggestion
Sizes and offsets have the identical bit width and both 32-bit and 64-bit
signed integer options are supported.

As in the List layout, the offsets encode the start position of each slot
in the child array. In contrast to the List layout, list lengths are stored
explicitly in the sizes buffer instead of inferred. This allows offsets to be
```

out of order. Elements of the child array do not have to be stored in the
same order they logically appear in the list elements of the parent array.

When a value is null, the corresponding offset and size can have arbitrary
values. When size is 0, the corresponding offset can have an arbitrary value.
If choosing a value is possible, we recommend setting offsets and sizes to 0 in
these cases.

A list-view type is specified like ``ListView<T>``, where ``T`` is any type
(primitive or nested). In these examples we use 32-bit offsets where
the 64-bit offset version would be denoted by ``LargeListView<T>``.

> **Example Layout: ``List<Int8>`` Array**

Reviewer 2: ```suggestion
**Example Layout: ``ListView<Int8>`` Array**
```

We illustrate an example of ``ListView<Int8>`` with length 4 having values::

    [[12, -7, 25], null, [0, -127, 127, 50], []]

will have the following representation: ::

    * Length: 4, Null count: 1
    * Validity bitmap buffer:

      | Byte 0 (validity bitmap) | Bytes 1-63            |
      |--------------------------|-----------------------|
      | 00001101                 | 0 (padding)           |

    * Offsets buffer (int32)

      | Bytes 0-3  | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-63           |
      |------------|-------------|-------------|-------------|-----------------------|
      | 0          | unspecified | 3           | unspecified | unspecified (padding) |

    * Sizes buffer (int32)

      | Bytes 0-3  | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-63           |
      |------------|-------------|-------------|-------------|-----------------------|
      | 3          | unspecified | 4           | 0           | unspecified (padding) |

>     * Values array (Int8array):

Reviewer 2: ```suggestion
    * Values array (Int8Array):
```

          |-------------------------------|-----------------------|
          | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 | unspecified (padding) |

ListView Layout
~~~~~~~~~~~~~~~

The ListView layout is defined by three buffers: a validity bitmap, an offsets
buffer, and an additional sizes buffer. Sizes and offsets have the identical bit
width and both 32-bit and 64-bit signed integer options are supported.

As in the List layout, the offsets encode the start position of each slot in the
child array. In contrast to the List layout, list lengths are stored explicitly
in the sizes buffer instead of inferred. This allows offsets to be out of order.
Elements of the child array do not have to be stored in the same order they
logically appear in the list elements of the parent array.

When a value is null, the corresponding offset and size can have arbitrary

> values. When size is 0, the corresponding offset can have an arbitrary value.

Reviewer: Hmm... I wonder if we should be stricter here and accept only values that fall within the bounds of the child array. It might make life easier for language-strict implementations, and perhaps specific instrumentations such as ASan / UBsan.

My question would be: is it detrimental in any way to be stricter? Intuitively, it doesn't seem performance would suffer (as populating a ListView array would typically keep a running offset, valid by construction).

Author: My intention here is to say that an implementation that trusts `offsets[i]` when `sizes[i] == 0` is non-compliant and unsafe. I consider that being more strict on what consumers of list-views can do than saying "go ahead and feel free to dereference any offset". Being stricter on producers of list-views doesn't make the consumers protected from malicious or bogus producers.

> ...and perhaps specific instrumentations such as ASan / UBsan.

Being able to pass random list-view arrays with wild `offsets[i]` when `sizes[i] == 0` is how we can fuzzy the implementations and get ASan / UBSan to catch the mistakes. Well-behaved `offsets[i]` will more likely hide issues than highlight them.

Reviewer: > I consider that being more strict on what consumers of list-views can do than saying "go ahead and feel free to dereference any offset". Being stricter on producers of list-views doesn't make the consumers protected from malicious or bogus producers.

That's a good point. But it also means that, instead of delegating handling of "wild" offsets to a validation routine, consumers have to be careful to accept such offsets at any place where ListView data is processed. This requires more care from implementations, with no upside that I can think of.

> Being able to pass random list-view arrays with wild offsets[i] when sizes[i] == 0 is how we can fuzz the implementations and get ASan / UBSan to catch the mistakes.

We already fuzz Arrow C++ with invalid data using the IPC layer. The way it's done is precisely to call the validation routine after having read the data (if reading the data was at all successful).

Conversely, we cannot possibly fuzz all internal routines (such as compute functions, etc.) against invalid or non-conventional data. We should indeed try to generate interesting (but valid) cases in our random generation functions, but not all internal routines are tested against random data.

So it seems to me that being stricter at the edge helps safety more than making the format more lenient so that we can better fuzz the internals.

@alamb @tustvold From the perspective of the Arrow Rust implementation, what is your opinion?

Reviewer 3: From the perspective of the Rust implementation we will almost certainly require all views to be valid, this makes it significantly simpler to provide safe APIs, that can separate null mask handling from value access.

Author: Alright, allowing for any value that is within-bounds sounds like a viable compromise here.

Author: > Hmm... I wonder if we should be stricter here and accept only values that fall within the bounds of the child array.

@pitrou I think it could be challenging to forbid the `child_data[0].length()` offset altogether. It's the natural value to have when the last list view appended is an `[]`.

Author: And for the case where the entire list-view is just `[[]]`. `offsets[0] == child_data[0].length() == 0`.

Reviewer: Yes, I think it's ok to point to the first byte past the last if the length is 0 (similar to an `end()` iterator in C++).

Reviewer 3: >  Being stricter on producers of list-views doesn't make the consumers protected from malicious or bogus producers.

> So it seems to me that being stricter at the edge helps safety more than making the format more lenient

FWIW this is the approach of arrow-rs, and Rust in general, all data is either validated or constructed in such a way as to statically guarantee the necessary invariants.

> If choosing a value is possible, we recommend setting offsets and sizes to 0 in

Reviewer: I'm curious why there would be situations where picking a value is not possible?

Author: It's not so much about possibility (so I'm very open to re-wording), but more about machine affinity (having less branches in loops, SIMD...). I can't predict all the possible ways people might want to save time in compute kernels.

For instance, when I slice the list-views for concatenation I have to subtract values from offsets (because I'm compressing the child array). I can do it in multiple ways:

 - `offsets[i] = offsets[i] - displacement;` leading to the possibility of negative offsets when `sizes[i] == 0` and `offsets[i]` is `unspecified` and potentially smaller than the positive displacement.
 - `offsets[i] = std::max(0, offsets[i] - displacement);` guaranteeing it's zero but introducing branching `*`
 - skipping setting `offsets[i]` at all when `sizes[i] == 0` introducing branching
 
I don't think this is an important enough operation but it illustrates how easy these stricter requirements could get in the way of micro optimizations. I suspect these become even more relevant when attempts to vectorize these loops are made.

`*` or requiring branchless saturating math operations

Reviewer 3: In my experience forcing kernels to handle the null mask often results in significantly less performant code, especially in the absence of hand-rolled SIMD, I suspect this will outweigh any potential benefits from allowing arbitrary values to be left at null positions.

In your example, surely you have to perform the bounds checks anyway as the non-null elements might be shorted than displacement?

Author: > In your example, surely you have to perform the bounds checks anyway as the non-null elements might be shorted than displacement?

`displacement` is greater than `0` but `<=` all the non-null and non-empty list-view offsets in the buffer. It's calculated by finding the smallest valid offset on the source array before applying the transformation. Finding the smallest requires checking null mask and `sizes[i]` buffer, applying the transformation is where many alternatives are available.

I will check the C++ implementation again, but in the Go implementation I'm doing all the checks.

Being more strict on the spec is more likely to constrain kernels writing/transforming list-views, but hopefully we compensate for that in efficiency of kernels reading list-views.

Author: In systems that also support selection vectors and not only validity bitmaps, skipping a pass over the values becomes a powerful optimization. Even more important than being able to blindly read or transform values at the null slots.

https://dl.acm.org/doi/abs/10.1145/3465998.3466009

Reviewer 3: Agreed, many kernels in arrow-rs do still handle the null mask, or some similar mask derived from the inputs, but make use CLZ tricks to iterate it efficiently. This in turn requires providing an API for value access, that doesn't itself then check the null mask, and in order for this to be safe we need the invariants to hold for non-null values :smile:

these cases.

A list-view type is specified like ``ListView<T>``, where ``T`` is any type

> (primitive or nested). In these examples we use 32-bit offsets where
> the 64-bit offset version would be denoted by ``LargeListView<T>``.

Reviewer 4: To be clear, the bitwidth applies to both offsets and sizes, right?
```suggestion
(primitive or nested). In these examples we use 32-bit offsets and sizes where
the 64-bit version would be denoted by ``LargeListView<T>``.
```

Author: Yes. I will reinforce it here as well.

**Example Layout: ``ListView<Int8>`` Array**

We illustrate an example of ``ListView<Int8>`` with length 4 having values::

    [[12, -7, 25], null, [0, -127, 127, 50], []]

> will have the following representation: ::

Reviewer: ```suggestion
It will have the following representation: ::
```

ListView Layout
~~~~~~~~~~~~~~~

The ListView layout is defined by three buffers: a validity bitmap, an offsets
buffer, and an additional sizes buffer. Sizes and offsets have the identical bit
width and both 32-bit and 64-bit signed integer options are supported.

As in the List layout, the offsets encode the start position of each slot in the
child array. In contrast to the List layout, list lengths are stored explicitly
in the sizes buffer instead of inferred. This allows offsets to be out of order.
Elements of the child array do not have to be stored in the same order they
logically appear in the list elements of the parent array.

Every list-view value, including null values, has to guarantee the following
invariants: ::

    0 <= offsets[i] <= length of the child array
    0 <= offsets[i] + size[i] <= length of the child array

A list-view type is specified like ``ListView<T>``, where ``T`` is any type
(primitive or nested). In these examples we use 32-bit offsets and sizes where
the 64-bit version would be denoted by ``LargeListView<T>``.

**Example Layout: ``ListView<Int8>`` Array**

We illustrate an example of ``ListView<Int8>`` with length 4 having values::

>     [[12, -7, 25], null, [0, -127, 127, 50], []]

Reviewer 5: It would be nice to have an example that shows what happens with duplicate lists and duplicate values:

```
[[12, -7, 25], null, [0, -127, 127, 12], [], [12, -7, 25]]
```

Author: @adriangb anything can happen: they can be duplicated in the data or entries can point to the same data.

Compact representation:

```
buffers:
  offsets: [0, _, 3, _, 0]
  sizes:   [3, _, 4, 0, 3]

children:
  values: [12, -7, 25, 0, -127, 127, 12]
```

Common representation:

```
buffers:
  offsets: [0, _, 3, _, 7]
  sizes:   [3, _, 4, 0, 3]

children:
  values: [12, -7, 25, 0, -127, 127, 12, 12, -7, 25]
```

Reviewer 5: Makes total sense, thanks!

It may have the following representation: ::

    * Length: 4, Null count: 1
    * Validity bitmap buffer:

      | Byte 0 (validity bitmap) | Bytes 1-63            |
      |--------------------------|-----------------------|
      | 00001101                 | 0 (padding)           |

    * Offsets buffer (int32)

      | Bytes 0-3  | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-63           |

>       |------------|-------------|-------------|-------------|-----------------------|

Author: Yes. I tried to make it clear that you can't expect sizes or offsets to be zero on NULL lists. The only constraint is that offset+size < children[0].length

>       |------------|-------------|-------------|-------------|-----------------------|

Reviewer 6: @felipecrv thanks for the reply!

bitmap. However, when treated independently, corresponding entries of the
children array will be non-null.

Union Layouts

> -------------

Reviewer 4: If you want to fix one more issue in this document, there seems to be an issue with the `Variadic Buffers` header:

https://github.com/apache/arrow/actions/runs/6381270929/job/17317395755?pr=37877#step:6:5783

Author: Fixed.
